File: experiments/016-LanguageComparisonRepeated/cells/java-gradle_1-Empty_free/2019-2c-parcial2-masked/20260917-052505-32178/workdir/src/main/java/surveycrawler/GuardedRun.java
package surveycrawler;

import java.util.ArrayList;
import java.util.List;

public class GuardedRun extends CrawlerRunMode {

    private final List<Runnable> undoActions = new ArrayList<>();

    // guarded run

    @Override
    public void startGuardedRun(SurveyCrawler aCrawler) {
        aCrawler.signalNestedGuardedRun();
    }

    @Override
    public void finishGuardedRun(SurveyCrawler aCrawler) {
        aCrawler.runNormally();
    }

    @Override
    public void assertFinished(SurveyCrawler aCrawler) {
        aCrawler.signalUnfinishedGuardedRun();
    }

    // moving

    @Override
    public void slideTowards(SurveyCrawler aCrawler, Point aDirection) {
        aCrawler.signalSiltFoundDuringGuardedRun();
    }

    // undoing

    @Override
    public void recordUndo(Runnable anUndoAction) {
        undoActions.add(anUndoAction);
    }

    @Override
    public void recoverFromError(SurveyCrawler aCrawler) {
        for (int i = undoActions.size() - 1; i >= 0; i--) {
            undoActions.get(i).run();
        }
        aCrawler.runNormally();
    }
}

File: experiments/020-PythonVsJava/cells/python-pytest_1-Empty_free/2019-2c-parcial2-masked/20260921-212215-18237/output/project/src/surveycrawler/guarded_run.py
from surveycrawler.crawler_run import CrawlerRun
from surveycrawler.free_run import FreeRun


class GuardedRun(CrawlerRun):

    # initialization

    def __init__(self):
        self._undo_actions = []

    # undoing

    def register_undo(self, an_undo_action):
        self._undo_actions.append(an_undo_action)

    def undo_all(self):
        for an_undo_action in reversed(self._undo_actions):
            an_undo_action()

    # guarded run

    def start_guarded_run(self, a_crawler):
        a_crawler.signal_guarded_run_inside_guarded_run()

    def end_guarded_run(self, a_crawler):
        return FreeRun()

    def finish(self, a_crawler):
        a_crawler.signal_unfinished_guarded_run()

    # error handling

    def handle_error(self, a_crawler):
        a_crawler.change_to_free_run()
        self.undo_all()

    # moving

    def slide_crawler_to(self, a_crawler, a_position):
        a_crawler.signal_silt_found()
